src/docfx/lib/schema/JsonSchemaTransformer.cs

```diff
@@ -13,136 +14,146 @@ internal class JsonSchemaTransformer
+                        if (schema.Properties.TryGetValue(key, out var propertySchema))
                         {
-                            return Transform(file, context, propertySchema, value, errors);
-                        }, LazyThreadSafetyMode.PublicationOnly);
+                            if (action?.Invoke(propertySchema, key, value) ?? false)
+                                continue;
+
+                            TraverseXref(propertySchema, value);
+                        }
                     }
                 }
-                return dict;
             }
         }
 
-        private JToken Transform(Document file, Context context, JsonSchema schema, JToken token, List<Error> errors)
+        private JToken TransformToken(Document file, Context context, JsonSchema schema, JToken token, List<Error> errors)
         {
-            return Traverse(schema, token, (subSchema, node) =>
-            {
-                if (node.Type == JTokenType.Array || node.Type == JTokenType.Object)
-                {
-                    // don't support array/object transform now
-                    return (default, node);
-                }
-
-                var transformedScalar = TransformScalar(subSchema, file, context, node as JValue, errors);
-                return (default, transformedScalar);
-            });
-        }
-
-        private JToken Traverse(JsonSchema schema, JToken token, Func<JsonSchema, JToken, (string[], JToken)> transform)
-        {
-            schema = _definitions.GetDefinition(schema);
-            if (schema == null)
-            {
-                return token;
-            }
-
-            var (transformedKeys, transformedToken) = transform(schema, token);
-            transformedKeys = transformedKeys ?? Array.Empty<string>();
+            return TransformToken(schema, token);
 
-            switch (transformedToken)
+            JToken TransformToken(JsonSchema subSchema, JToken node)
             {
-                case JValue scalar:
-
-                    return scalar;
-
-                case JArray array:
-
-                    if (schema.Items == null || schema.ContentType != JsonSchemaContentType.None)
-                        return array;
-
-                    var newArray = new JArray();
-                    foreach (var item in array)
-                    {
-                        var newItem = Traverse(schema.Items, item, transform);
-                        newArray.Add(newItem);
-                    }
+                subSchema = _definitions.GetDefinition(subSchema);
 
-                    return newArray;
+                if (subSchema == null)
+                {
+                    return node;
+                }
 
-                case JObject obj:
+                return _transformedProperties.GetOrAdd(node, _ => Transform());
 
-                    var newObject = new JObject();
-                    foreach (var (key, value) in obj)
+                JToken Transform()
```

Comment:
consider _not_ use nested function here because it is hard to tell what variables it captures. the second parameter to `dictionary.GetOrAdd` needs to be as pure as possible.

In this case, `errors` is captured and used inside `Transform` function, when `TransformScalar` generates an error, the error will be added to the list of the first calls `TransformToken`, other calls to `TransformToken` will not have correct errors.

Reply:
good point, will change that